### PR TITLE
[rush] Rebuild ignores existing build cache entries

### LINKED ISSUE #2702: [rush] "rush rebuild" should work the same with buildCacheEnabled=true
## Summary

When the experimental build cache feature is enabled, the semantics of the `rush rebuild` command are a bit counterintutive:

- `buildCacheEnabled: false` - traditionally, `rush rebuild` forces all projects to get rebuilt, bypassing the incremental build logic
- `buildCacheEnabled: true` - in this mode, `rush rebuild` will extract tarballs from the cache if possible; it only forces a rebuild for projects that selectively do not use the cache.  You must use `rush rebuild --disable-build-cache` to get the old behavior.

  

## Details

I chatted with @iclanton about this, and we decided that it would make more sense to maintain the old semantics.

Proposed new design:
-  `rush rebuild` will always rebuild each project, regardless of `buildCacheEnabled`
- `rush rebuild` will populate the cache as usual (the cache is not "disabled", just skipped)
- The `--disable-build-cache` CLI parameter will be removed entirely

### PR DESCRIPTION
**Pull request by elliot-nelson**
## Summary

When a user runs `rush rebuild`, this should build everything whether we'd normally skip or retrieve from the build cache.  (Previously, rebuild was pulling from the build cache if an entry existed.)

Fixes #2702.

> Reviewer note: reviewing with [white space turned off](https://github.com/microsoft/rushstack/pull/2802/files?w=1) makes the diff much easier to see, since a large chunk of code was de-dented one level.

## Details

Running `rush rebuild` now rebuilds everything, regardless of incremental build state or build cache entries.

A few additional tweaks/clarifications:

 - If the build cache exists, and write is enabled, `rush rebuild` ignores cache entries but populates cache entries as it builds.  This seems like intuitive behavior.
 - `rush rebuild` still attempts to gather information about changed files and projects (which would be necessary to write out the build cache entry later).  However, if it encounters an unexpected error -- for example, maybe `git` is not initialized in the monorepo -- the usual warning is now suppressed.
 - The `--disable-build-cache` flag has been removed for all commands.  (The primary usage of this flag was to ignore the build cache while building, which is now easily done by switching from `rush build` to `rush rebuild`.)  For advanced fine-grained control over the build cache, the environment variables `RUSH_BUILD_CACHE_ENABLED` and `RUSH_BUILD_CACHE_WRITE_ALLOWED` can still be used.

## How it was tested

Tested locally using local rush:

 - `node apps/rush-lib/lib/start.js build --to @rushstack/eslint-plugin`
   - Targeted projects pulled from build cache
 - `node apps/rush-lib/lib/start.js rebuild --to @rushstack/eslint-plugin`
   - Targeted projects rebuild and populate the local build cache
 - Confirmed that if `.git` is renamed to `.xyz`, running the commands above produces expected behavior.
   - `rush build` generates a warning line that it is running a full rebuild
   - `rush rebuild` generates no warning line

## Remaining questions

The only thing I'm not 100% on is that warning line... for now I've left the verbiage alone and suppressed it for `rush rebuild`, but I don't know if that's entirely appropriate.  Having no git folder available, for example, does affect rebuild -- it prevents it from writing to the build cache if the build cache is enabled.

Possibly a better solution is to make the warning more generic ("Cannot retrieve or save any incremental or build cache data for this project.").

### CODE REVIEW

**Comment by octogonz** [comment]
> The only thing I'm not 100% on is that warning line... for now I've left the verbiage alone and suppressed it for `rush rebuild`, but I don't know if that's entirely appropriate. Having no git folder available, for example, does affect rebuild -- it prevents it from writing to the build cache if the build cache is enabled.

This seems okay to me.  An everyday user might be puzzled why `rush build` is super slow, so the warning is useful there.  An everyday user, working outside of Git, who explicitly chose `rush rebuild`, probably does not care about populating the cache.

> Possibly a better solution is to make the warning more generic ("Cannot retrieve or save any incremental or build cache data for this project.").

Showing something like this when the `--debug` option is specified could be helpful.

File: common/changes/@microsoft/rush/rebuild_2021-07-09-20-31.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "rush rebuild ignores existing build cache entries",
```
Comment (octogonz) [inline common/changes/@microsoft/rush/rebuild_2021-07-09-20-31.json:5]:
```suggestion
      "comment": "When the experimental build cache is enabled, \"rush rebuild\" now forces cached projects to be rebuilt  (GitHub #2802)",
```

File: common/changes/@microsoft/rush/rebuild_2021-07-09-20-31.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+      "type": "none"
+    }
```
Comment (octogonz) [inline common/changes/@microsoft/rush/rebuild_2021-07-09-20-31.json:7]: Let's also add a second change log that is like:
```
"(Breaking change) Remove the experimental \"--disable-build-cache\" command line parameter."
```

Although the feature is still experimental, these are nontrivial changes, so we should remember to publish them as a MINOR bump.

File: common/changes/@microsoft/rush/rebuild_2021-07-13-21-32.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "(Breaking change) Remove the experimental \"--disable-build-cche\" command line parameter.",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/rebuild_2021-07-13-21-32.json:5]:
```suggestion
      "comment": "(Breaking change) Remove the experimental \"--disable-build-cache\" command line parameter.",
```

**Comment by iclanton** [comment]
>The only thing I'm not 100% on is that warning line... for now I've left the verbiage alone and suppressed it for rush rebuild, but I don't know if that's entirely appropriate. Having no git folder available, for example, does affect rebuild -- it prevents it from writing to the build cache if the build cache is enabled.

>Possibly a better solution is to make the warning more generic ("Cannot retrieve or save any incremental or build cache data for this project.").

I think some warning should be printed. I'd expand your proposed phrasing to "Cannot retrieve or save any incremental or build cache data for this project. This may be because the project is not in a Git repo or Git is not available."

Maybe also mention something about "This can be suppressed by setting XYZ environment variable to somesuch value".

**Comment by iclanton** [comment]
@octogonz - Should we consider disabling the incremental build when the build cache is enabled?

File: apps/rush-lib/src/logic/taskRunner/ProjectBuilder.ts [hunk 4]
```diff
@@ -227,15 +227,17 @@ export class ProjectBuilder extends BaseBuilder {
...
-          'Error calculating incremental build state. Instead running full rebuild. ' + error.toString()
-        );
+        if (this.isIncrementalBuildAllowed) {
+          terminal.writeLine(
+            'Error calculating incremental build state. Instead running full rebuild. ' + error.toString()
+          );
+        }
       }
 
       const isPackageUnchanged: boolean = !!(
```
Comment (iclanton) [inline apps/rush-lib/src/logic/taskRunner/ProjectBuilder.ts:243]: This variable is only used inside the `if (this.isIncrementalBuildAllowed) { ... }` block. Can this get moved into that block? Maybe even only after we try to restore from cache?

**Comment by octogonz** [comment]
> I think some warning should be printed. I'd expand your proposed phrasing to "Cannot retrieve or save any incremental or build cache data for this project. This may be because the project is not in a Git repo or Git is not available."
> 
> Maybe also mention something about "This can be suppressed by setting XYZ environment variable to somesuch value".

If you want to print a warning, maybe it could be more broad disclaimer like:

```
This folder is not a Git repository. Rush depends on Git for certain features such as 
incremental builds, caching, and change detection. Those features will be disabled.
```

What we should avoid is a wall of warning text that is usually irrelevant to what the user is thinking about. A long, accurate, thorough notice is useless in a context where nobody will bother to read it. 😉

**Comment by iclanton** [comment]
There is an important difference between "not tracked by git" and "the git binary isn't present." If the warning says the former, but the issue is the latter, that'll mislead whoever is trying to troubleshoot.

**Comment by octogonz** [comment]
> Should we consider disabling the incremental build when the build cache is enabled?

@iclanton  This is an interesting idea and probably a simple change.

But I feel like it should happen in a separate MINOR bump of Rush, because it affects a bunch of edge cases that are difficult to test, and are normally invisible to end users. If it breaks something and people need to temporarily downgrade, we wouldn't want to revert Elliot's PR (which is a breaking design change and harder to revert).

**Comment by elliot-nelson** [comment]
Ready for re-review - incorporated feedback and updated two warning messages.

#### (1) If _git is unavailable or not working_ (the git binary is returning non-zero exit codes)

Old Message:

```text
Unable to calculate incremental build state. Instead running full rebuild. Ensure Git is present.
```

New Message:

```text
Warning: incremental builds, caching, and change detection are disabled.
Ensure this workspace is tracked by git and git is available.
```

#### (2) If some other unexpected error occurs while collecting file hashes

Old Message:

```text
Error calculating incremental build state. Instead running full rebuild. [error]
```

New Message:

```text
Error encountered calculating incremental build state: [error]
Incremental builds, caching, and change detection are disabled.
```

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
Released with Rush 5.50.0
